src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+                    currentState = Interlocked.CompareExchange(ref _state, knownState + LockStates.WaitingReader, knownState);
+                    if (currentState != knownState)
+                    {
+                        continue;
+                    }
+
+#if RWLOCK_STATISTICS
+                    Interlocked.Increment(ref _readerContentionCount);
+#endif
+
+                    int modifyState = -LockStates.WaitingReader;
+                    ManualResetEvent readerEvent = null;
+                    bool waitSucceeded = false;
+                    try
+                    {
+                        readerEvent = GetOrCreateReaderEvent();
+                        waitSucceeded = readerEvent.WaitOne(millisecondsTimeout);
+
+                        // AcquireReaderLock cannot have reentry via pumping while waiting for readerEvent, so
+                        // lockEntry's state should not change from underneath us
+                        Debug.Assert(lockEntry.HasLockID(_lockID));
+
+                        if (waitSucceeded)
+                        {
+                            // Become a reader
+                            Debug.Assert((_state & LockStates.ReaderSignaled) != 0);
+                            Debug.Assert((_state & LockStates.ReadersMask) < LockStates.ReadersMask);
+                            modifyState += LockStates.Reader;
+                        }
+                    }
+                    finally
+                    {
+                        // Make the state changes determined above
+                        knownState = Interlocked.Add(ref _state, modifyState) - modifyState;
+
+                        if (!waitSucceeded)
+                        {
+                            // Check for last signaled waiting reader
+                            if ((knownState & LockStates.ReaderSignaled) != 0 &&
+                                (knownState & LockStates.WaitingReadersMask) == LockStates.WaitingReader)
+                            {
+                                if (readerEvent == null)
+                                {
+                                    readerEvent = _readerEvent;
+                                    Debug.Assert(readerEvent != null);
+                                }
+
+                                // Ensure the event is signalled before resetting it, since the ReaderSignaled state is set
+                                // before the event is set.
+                                readerEvent.WaitOne();
+                                Debug.Assert((_state & LockStates.ReadersMask) < LockStates.ReadersMask);
+
+                                // Reset the event and lower reader signaled flag
+                                readerEvent.Reset();
+                                Interlocked.Add(ref _state, LockStates.Reader - LockStates.ReaderSignaled);
+
+                                // Honor the orginal status
+                                ++lockEntry._readerLevel;
```

Comment:
The lock entry is thread-local, and its _readerLevel indicates the number read locks taken by the thread. This is a rare case where the wait timed out, but shortly afterwards, waiting readers got released, hence the ReaderSignaled bit is set. While the ReaderSignaled bit is set, new requests for a write lock must spin or wait to acquire the lock, so it is safe for this thread to acquire a read lock and call ReleaseReaderLock() as a shortcut to do the work of releasing other waiters.